Add abandon changes popup to data driven forms components

rvsia opened the pull request:
Part of https://bugzilla.redhat.com/show_bug.cgi?id=1725927

**Description**

DDF components are missing a option to tell the Abandon changes popup to show up. This PR connects all Data driven forms to redux and store pristine of the form in its store.

The same thing should be done for Tag component.

**Before**

![image](https://user-images.githubusercontent.com/32869456/64795218-05b73880-d57e-11e9-886c-54cff5de12e0.png)

**After**

![image](https://user-images.githubusercontent.com/32869456/64795041-bd981600-d57d-11e9-9443-e08b4a63ba8e.png)

@miq-bot add_label bug, ivanchuk/yes, react

rvsia commented on config/jest.setup.js:
```diff
@@ -31,3 +31,9 @@ document.body.createTextRange = () => ({
+
+import configureStore from 'redux-mock-store';
+
+ManageIQ.redux.store = configureStore([])({});
```
I replaced it with the real store.

Hyperkid123 commented on app/javascript/forms/data-driven-form.jsx:
```diff
@@ -27,6 +33,7 @@ const MiqFormRenderer = ({ className, ...props }) => (
       'pristine',
       'invalid',
     ]}
+    onStateUpdate={({ pristine }) => setPristine(pristine)}
```
This is not going to work if somebody sends `onStateUpdate` (its being used). You should merge this with the callback from props.

rvsia commented:
@miq-bot add_label ivanchuk/no
@miq-bot remove_label ivanchuk/yes

Will need to create a separate Ivanchuk PR because of different test set.

himdel commented:
@rvsia looks like this should be 2 different PRs, most of the mount/shallow/deep/children changes seem completely unrelated to the main issue.

And we may want to expose the whole thing instead of just `pristine` - what about exposing both `pristine` and `valid` (is there a way to get a list of validation errors?) 

(In fact, if the plan is still to use FormButtons eventually, we would need to expose `newRecord`, `saveable` (=valid) and `pristine`.)

Also, one question - what clears `formReducer` from the state? If you're in an explorer controller and use ajax to replace main_div, will any subsequent navigation fail because `formReducer` still exists?

Hyperkid123 commented:
> if the plan is still to use FormButtons 

Personally I did rethink usage FormButtons. I don't really see a benefit of this component (for data-driven-forms) when these buttons are provided by the library.

That said adding more or the form state to redux store is of course possible and we can definitely add it if its required. `onStateUpate` returns the whole formState with values, errors and other meta data so if we end up with the generic buttons solution we are OK. But i think we should add it when we actually need it.

> Also, one question - what clears formReducer from the state? If you're in an explorer controller and use ajax to replace main_div, will any subsequent navigation fail because formReducer still exists?

I do agree that there should be reset dispatch call. Reseting might be done when `main_div` is replaced. That would be probably the safest place to do it.

rvsia commented:
@himdel 

ad tests) `connect` adds a new component to the tests' htmls, so all tests which were using shallow (renders only the first component) and tests with children (because there is one more now) fails. I could do more PRs like: 'connect DDF to redux', 'expose DDF forms to redux', 'check DDF redux changes for abandon popup'... but I think it would take too much time to get it merge and then backport it to Ivanchuk.

ad exposing) I want to expose only values which are needed right now. If we need any changes in the future, let's change it when we need it.

ad ajax rendering) Great point! I will investigate how it behaves on explorer screens.

_written before Martin M :smile_cat:_

himdel commented:
OK, IMO exposing both pristine and valid makes sense if we're exposing one of them, but.. you're right that we don't need it now, won't block on this  :+1: :)

Ok, so, sounds ilke this could be ready, let me know what you find about the explorers please :)

rvsia commented:
@himdel I added a function which clears the `formReducer` every time `ManageIQ.explorer` renders maindiv/right_cell. Also, added resetting tagging values after unmounting the component. Right now its a sufficient solution. However, I will add a proper reset action to the tag reducer, but that has to be done in the React-UI components package and I don't want to block this PR.

rvsia commented:
@miq-bot remove_label unmergeable

himdel commented on app/assets/javascripts/miq_explorer.js:
```diff
@@ -341,5 +347,7 @@ ManageIQ.explorer.processReplaceRightCell = function(data) {
     miqSparkleOff();
   }
 
+  ManageIQ.explorer.resetReduxForms();
```
@rvsia This is not the right place..  `ManageIQ.explorer.process` is the main entrypoint, `processReplaceRightCell` is only one possible branch.. `processFlash` and `processReplaceMainDiv` also handle partial updates.

(But you mentioned you'll deal with this more in a follow-up PR, and this won't break anything as is, so ignoring.)

rvsia replied:
`resetReduxForms` is also in `processReplaceMainDiv`. I didn't want to add this function to `processFlash` as I thought that it manages the flash messages (or does it do something else?) and I don't want to reset the forms when messages change. Right now, all forms are in `main_div` or in `right_cell`, so it is a sufficient (and I suppose all forms would be there).

himdel replied:
Aaah, sorry, I missed it in processReplaceMainDiv, you're right :)

himdel commented:
Old forms still work as expected :+1: 
Angular forms using `$scope` work as expected :+1: 
(Angular forms using `vm` are broken on master, adressing in ...)

React forms work as expected :+1: 
:tada:

himdel commented:
@rvsia Could this be ivanchuk/yes?

I'm basing my angular forms miqCheckForChanges fix on this PR and that fix should definitely land in ivanchuk.

rvsia commented:
@himdel No :( because Ivanchuk has different test files for react forms (it is missing newer forms), so I need to prepare another PR. But the code should be the same.

himdel commented:
That doesn't actually sound like a no, if it's just missing tests, Satoe can usually handle these issues :).
=> Thanks, adding ivanchuk/yes.

rvsia commented:
@himdel Oh, didn't know, perfect! :)

Hyperkid123 commented:
:haircut_man:

romanblanco commented:
@rvsia after merging this PR, [changes](https://github.com/ManageIQ/manageiq-ui-classic/pull/6184/files#diff-23ab9acb538585d8324a9f4d1497e5a5R314-R319) force user to answer `"Abandon changes?"` question every time they switch accordion in Configuration (Ops) page, even though the user is not if a form doing any changes he could abandon.

rvsia commented:
@romanblanco Testes on master, everything seems OK. There was an additional fix in https://github.com/ManageIQ/manageiq-ui-classic/pull/6223

romanblanco commented:
@rvsia you're right, all works now :+1:
